feat: Add new datatype JSON which is supported by Postgres and MariaDB

sunilgoli opened the pull request:
JSON data type is supported by both Postgres and MariaDB. 

add new file json.js (which contains control definition) under public/js/frappe/form/controls
Update control.js to import json.js.
Updated database.py under database/mariadb and database/mariadb with JSON in setup_type_map().
Added JSON in DocFields.

Removed:
The initial commit was with JSONB and JSON but later removed because of JSONB is not supported in MariaDB. Adding JSONB only to Postgres involved setup code which is used during the site setup.

sunilgoli committed:
Added 2 new datatypes for Postgres

Postgres supports JSON and JSONB data type

sunilgoli committed:
Removed JSONB datatype.

JSONB is only for Postgres.
To add this it requires code change in site setup/installation  process.

sunilgoli committed:
Removed JSONB from DocField

sunilgoli committed:
JS syntax fix

added a space after if

hrwX commented:
@Kratos-Freyja Any use-case for this ?

sahil28297 commented:
Rebasing it to develop

sunilgoli committed:
[BUG FIX] JSON datatype bug fix.

New datatype was not added in frappe.model.__init__.py file under data_fieldtypes list which was not allowing the json fields to be saved or retrieved

netchampfaris commented on frappe/public/js/frappe/form/controls/json.js:
```diff
@@ -0,0 +1,29 @@
+frappe.ui.form.ControlJSON = frappe.ui.form.ControlData.extend({
```
It's a better idea to extend `ControlCode` as it will support JSON syntax highlighting.

Kratos-Freyja replied:
Thanks, It's updated.

Kratos-Freyja replied:
By changing to ControlCode from ControlData it's raising a few issues If you can resolve this and change, please do.

sunilgoli replied:
Updated. ControlJSON now extends ControlCode

sunilgoli committed:
JSON component update

Building JSON component from ControlCode than ControlData

sunilgoli committed:
JSON component update

Rollback from ControlCode to ControlData due to ui issues

sunilgoli committed:
[Update] Update JSON base class to ControlCode

Update JSON base class to ControlCode from ControlData. Overridden set_language() to set this.df.options to JSON. Overridden set_formatted_input() to convert the value to string from json object.

surajshetty3416 commented on frappe/public/js/frappe/form/controls/json.js:
```diff
@@ -0,0 +1,15 @@
+frappe.ui.form.ControlJSON = frappe.ui.form.ControlCode.extend({
```
@Kratos-Freyja
1. Please add JSON validation. JSON control should check if the entered JSON is valid.
---
2. Please fix the following.
(It should be stringified value of JSON instead of [**object Object**])
<img width="983" alt="Screenshot 2019-10-22 at 7 52 35 PM" src="https://user-images.githubusercontent.com/13928957/67296166-b6154680-f505-11e9-8a65-3ff6784f80bb.png">
<img width="710" alt="Screenshot 2019-10-22 at 7 52 23 PM" src="https://user-images.githubusercontent.com/13928957/67296167-b6154680-f505-11e9-86c4-1afdda4851a2.png">
